src/lib.rs

```diff
@@ -280,35 +322,52 @@ impl<'a> Extract<'a> {
         let source = fs::File::open(self.source)?;
         let archive = self.archive.unwrap_or_else(|| detect_archive(&self.source));
 
-        match archive {
-            ArchiveKind::Plain(compression) | ArchiveKind::Tar(compression) => {
-                let mut reader = Self::get_archive_reader(source, compression);
-
-                match archive {
-                    ArchiveKind::Plain(_) => {
-                        match fs::create_dir_all(into_dir) {
-                            Ok(_) => (),
-                            Err(e) => {
-                                if e.kind() != io::ErrorKind::AlreadyExists {
-                                    return Err(Error::Io(e));
-                                }
+        // We cannot use a feature flag in a match arm. To bypass this the code block is
+        // isolated in a closure and called accordingly.
+        let extract_into_plain_or_tar = |source: fs::File, compression: Option<Compression>| {
+            let mut reader = Self::get_archive_reader(source, compression);
+
+            match archive {
+                ArchiveKind::Plain(_) => {
+                    match fs::create_dir_all(into_dir) {
+                        Ok(_) => (),
+                        Err(e) => {
+                            if e.kind() != io::ErrorKind::AlreadyExists {
+                                return Err(Error::Io(e));
                             }
                         }
-                        let file_name = self.source.file_name().ok_or_else(|| {
-                            Error::Update("Extractor source has no file-name".into())
-                        })?;
-                        let mut out_path = into_dir.join(file_name);
-                        out_path.set_extension("");
-                        let mut out_file = fs::File::create(&out_path)?;
-                        io::copy(&mut reader, &mut out_file)?;
-                    }
-                    ArchiveKind::Tar(_) => {
-                        let mut archive = tar::Archive::new(reader);
-                        archive.unpack(into_dir)?;
                     }
-                    _ => unreachable!(),
-                };
+                    let file_name = self
+                        .source
+                        .file_name()
+                        .ok_or_else(|| Error::Update("Extractor source has no file-name".into()))?;
+                    let mut out_path = into_dir.join(file_name);
+                    out_path.set_extension("");
+                    let mut out_file = fs::File::create(&out_path)?;
+                    io::copy(&mut reader, &mut out_file)?;
+                }
+                #[cfg(feature = "archive-tar")]
+                ArchiveKind::Tar(_) => {
+                    let mut archive = tar::Archive::new(reader);
+                    archive.unpack(into_dir)?;
+                }
+                #[allow(unreachable_patterns)]
+                _ => unreachable!(),
```

Comment:
can you add a comment here explaining that `detect_archive` will blow up for file formats that aren't supported before it gets to this point?

Reply:
I did the minimum changes so that the code would still compile. This means `detect_archive()` simply panics in case it cannot do its job. Wouldn't it be better if it simply returned a `Result` instead? Then no need for comments. That is a larger change though. But I'd say it is worth it. What do you think?

Reply:
Meanwhile I've unified the unreachables in 21825d0b737b4ebddc893fd8e63fb6fc.

Reply:
I agree, I think returning a `Result` would be better. Do you mind tackling that change in this PR?

Reply:
276a30973dd0b7a867d8fb99f2cde06a changes the `detect_archive()` API to return a `Result` instead of panicking.